### PR TITLE
proposal: add --keep to runc run

### LINKED ISSUE #2817: proposal: add --no-delete to runc run
`runc run` is much more simple to use than `create/start` async commands that are almost impossible to use safely without introducing a special shim-like binary to deal with `SIGCHLD` reaper issues.

One of the problems though is that when using `run` calling process can't get the full stats of the process because cgroups are already deleted when it returns. Could we consider adding `--no-delete` option that leaves cgroups and other resources in place so the caller can analyze them and then call `runc delete` to do the manual cleanup.

### PR DESCRIPTION
**Pull request by lifubang**
To close #2817

As described in #2817, for some times, we need to analyze the container's resource usage, so we expect runc does not delete the container after it exits.

Signed-off-by: lifubang <lifubang@acmcoder.com>

## Changelog entry
 (by @kolyshkin)
```
New features:
* runc run: new `--keep` option to skip removal exited containers artefacts.
  This might be useful to check the state (e.g. of cgroup controllers) after
  the container has￼exited. (#2817)
```

### CODE REVIEW

File: run.go
Comment (thaJeztah) [inline run.go:45]: small typo (s/destory/destroy/)☺️

```suggestion
			Usage: "don't auto-destroy the container after it stopped",
```

Or perhaps

```suggestion
			Usage: "do not delete the container after it exits",
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: utils_linux.go
Comment (thaJeztah) [inline utils_linux.go:444]: Not something new in this PR (so don't change) but it would be good to use the "default" values match the default behavior. (e.g. change the field to `keep` or `disableAutoDestroy`), so that the default for "boolean" values (`false`) also is the expected "default"

Comment (lifubang) [reply utils_linux.go:444]: Yes, it dose look strange to treat the default value `false` of `--no-delete` as auto-delete the container.
May be we can change `--no-delete` to `--disable-autodestroy`?

Comment (kolyshkin) [reply utils_linux.go:444]: I have suggested `--keep-cgroup` in #2817, but as we also have to keep the runc state directory, maybe just `--keep`?

File: tests/integration/run.bats
Comment (kolyshkin) [inline tests/integration/run.bats:14]: perhaps just `true`?

File: tests/integration/run.bats
Comment (kolyshkin) [inline tests/integration/run.bats:31]: Maybe also check that the cgroup is there?

File: utils_linux.go
Comment (thaJeztah) [reply utils_linux.go:444]: Yes, I was considering "--keep" as well; so sgtm

Comment (lifubang) [reply utils_linux.go]: SGTM too.
Welcome more better suggestions...

Comment (tonistiigi) [reply utils_linux.go]: The reason I suggested `--no-delete` was because it is consistent with the three existing `--no-*` flags in `runc run` and `delete` is consistent with the existing `runc delete` command that has the matching behavior. As a user, I don't care of course, and can use any flag.

Comment (kolyshkin) [reply utils_linux.go]: The existing `--no-*` flags are
```
   --no-subreaper            disable the use of the subreaper used to reap reparented processes
   --no-pivot                do not use pivot root to jail process inside rootfs.  This should be used whenever the rootfs is on top of a ramdisk
   --no-new-keyring          do not create a new session keyring for the container.  This will cause the container to inherit the calling processes session key
```

and in all 3 cases it's a `--no-NOUN`. There's also "--preserve-fds`...

To me `--keep` still looks like the best choice, the only issue is it's too generic (keep everything?).

File: tests/integration/run.bats
Comment (kolyshkin) [inline tests/integration/run.bats:48]: Maybe it's easier to check something like pids limit:
```
        check_cgroup_value "pids.max" max
```

File: tests/integration/run.bats
Comment (kolyshkin) [inline tests/integration/run.bats:43]: nit: `# check that cgroup exists`

File: tests/integration/run.bats
Comment (kolyshkin) [inline tests/integration/run.bats:34]: please remove this comment, it is of no use

**Comment by kolyshkin** [comment]
> Because of systemd's unit garbage collection policy, it seems that there is no way to keep container's cgroup paths after it exit when we use systemd driver.

Hmm, do we have a place in the documentation to add this fact? I think we need to create one (working on it)

**Comment by tonistiigi** [comment]
@kolyshkin What's the blocker in here?

File: utils_linux.go
Comment (tonistiigi) [reply utils_linux.go]: I guess I wouldn't consider `pivot` noun in this context (short for "pivot root"). Any flag is ok for me though.

Comment (cyphar) [reply utils_linux.go]: The difference is that for all of those `--no` settings, you're disabling a security feature and so the default should be false. Since the default for this option should also be false, it should be named something that makes sense in that context.

**Comment by AkihiroSuda** [comment]
The blocker was v1.0.0 GA, but it will be released _really_ soon, finally 😅

**Comment by kolyshkin** [comment]
Did another review round. As for the name for the option, I don't like `--keep` much, nor do I like `--no-delete`.

Since there are no better options, I guess either `--keep` or `--no-delete` is fine with me.

Setting the milestone to 1.1.0.

**Review by kolyshkin — COMMENTED** [review]
LGTM, except we need to describe the new option in https://github.com/opencontainers/runc/blob/master/man/runc-run.8.md

I propose
```markdown
**--keep**
: Keep container's state directory and cgroup. This can be helpful if a user wants to check
the state (of cgroup etc.) after the container has exited. Note if this option is used, a manual
**runc delete** is needed afterwards.
```

**Comment by kolyshkin** [comment]
I took a liberty to add a commit with a man page, and rebase.

CI is expected to fail because of #3050.

**Review by AkihiroSuda — APPROVED** [review]
LGTM if green

**Comment by kolyshkin** [comment]
Rebased to kick the latest CI. Added labels. Provided changelog entry. Let's merge this already.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
@cyphar PTAL

**Comment by AkihiroSuda** [comment]
@cyphar LGTY?

**Comment by AkihiroSuda** [comment]
> Merging is blocked
> Merging can be performed automatically once the requested changes are addressed.

Seems to need LGTM from @thaJeztah

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**
